src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -177,13 +177,46 @@ void CreatePrimaryRequest<I>::handle_refresh_image(int r) {
 
 template <typename I>
 void CreatePrimaryRequest<I>::unlink_peer() {
+  // TODO: Document semantics for unlink_peer
   uint64_t max_snapshots = m_image_ctx->config.template get_val<uint64_t>(
     "rbd_mirroring_max_mirroring_snapshots");
   ceph_assert(max_snapshots >= 3);
 
   std::string peer_uuid;
   uint64_t snap_id = CEPH_NOSNAP;
 
+  // If snap has no peers, remove snapshot. This can happen when mirror
+  // removes peer from peer_uuids when sync is completed and no other
+  // peers remain.
+  for (auto &peer : m_mirror_peer_uuids) {
+    std::shared_lock image_locker{m_image_ctx->image_lock};
+    for (auto &snap_it : m_image_ctx->snap_info) {
+      auto info = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
+        &snap_it.second.snap_namespace);
+      if (info == nullptr) {
+        continue;
+      }
+      if (info->mirror_peer_uuids.empty() || !info->complete) {
+        peer_uuid = peer;
+        snap_id = snap_it.first;
+        break;
+      }
+      if (snap_it.first == CEPH_NOSNAP) {
+        break;
+      }
+    }
+  }
+
+  if (snap_id != CEPH_NOSNAP) {
+    auto ctx = create_context_callback<
+      CreatePrimaryRequest<I>,
+      &CreatePrimaryRequest<I>::handle_unlink_peer>(this);
+    auto req = UnlinkPeerRequest<I>::create(m_image_ctx, snap_id, peer_uuid, true,
+                                            ctx);
+    req->send();
+    return;
```

Comment:
This duplicates `UnlinkPeerRequest` creation code while missing the log message.  Here is how I would do it (I incorporated all comments here):
```
template <typename I>
void CreatePrimaryRequest<I>::unlink_peer() {
  // TODO: Document semantics for unlink_peer
  uint64_t max_snapshots = m_image_ctx->config.template get_val<uint64_t>(
    "rbd_mirroring_max_mirroring_snapshots");
  ceph_assert(max_snapshots >= 3);

  std::string peer_uuid;
  uint64_t snap_id = CEPH_NOSNAP;

  {
    std::shared_lock image_locker{m_image_ctx->image_lock};
    for (auto &peer : m_mirror_peer_uuids) {
      for (auto &snap_it : m_image_ctx->snap_info) {
        auto info = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
          &snap_it.second.snap_namespace);
        if (info == nullptr || !info->is_primary()) {
          continue;
        }
        if (!info->mirror_peer_uuids.empty() &&
            info->mirror_peer_uuids.count(peer) == 0) {
          // snapshot is linked to other peer(s)
          continue;
        }
        if (info->mirror_peer_uuids.empty() || !info->complete) {
          peer_uuid = peer;
          snap_id = snap_it.first;
          goto do_unlink;
        }
      }
    }

    for (auto &peer : m_mirror_peer_uuids) {
      size_t count = 0;
      uint64_t unlink_snap_id = 0;
      for (auto &snap_it : m_image_ctx->snap_info) {
        auto info = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
          &snap_it.second.snap_namespace);
        if (info == nullptr) {
          continue;
        }
        if (info->state != cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
          // reset counters -- we count primary snapshots after the last promotion
          count = 0;
          unlink_snap_id = 0;
          continue;
        }
        if (info->mirror_peer_uuids.count(peer) == 0) {
          // snapshot is not linked with this peer
          continue;
        }
        count++;
        if (count == max_snapshots) {
          unlink_snap_id = snap_it.first;
        }
        if (count > max_snapshots) {
          peer_uuid = peer;
          snap_id = unlink_snap_id;
          goto do_unlink;
        }
      }
    }
  }

  finish(0);
  return;

do_unlink:
  CephContext *cct = m_image_ctx->cct;
  ldout(cct, 15) << "peer=" << peer_uuid << ", snap_id=" << snap_id << dendl;

  auto ctx = create_context_callback<
    CreatePrimaryRequest<I>,
    &CreatePrimaryRequest<I>::handle_unlink_peer>(this);
  auto req = UnlinkPeerRequest<I>::create(m_image_ctx, snap_id, peer_uuid, true,
                                          ctx);
  req->send();
}
```
Note the additional benefit of not dropping `m_image_ctx->image_lock` between iterations and thus getting a consistent view across both loops.

Reply:
Good deal, I've rechecked this to confirm all comments are incorporated. Since I cannot argue that this isn't cleaner than two calls to `UnlinkPeerRequest` I've used this code in the patch. Thanks.